src/linux/component.rs

```diff
@@ -1,160 +1,388 @@
-    let mut matchings: HashMap<u32, Vec<String>> = HashMap::with_capacity(10);
+/// Information about thermal sensor, may be unavailable it's
+/// kernel module and chip dependant.
+enum TermalSensorType {
+    /// 1: CPU embedded diode
+    CPUEmbeddedDiode,
+    /// 2: 3904 transistor
+    Transistor3904,
+    /// 3: thermal diode
+    ThermalDiode,
+    /// 4: thermistor
+    Thermistor,
+    /// 5: AMD AMDSI
+    AMDAMDSI,
+    /// 6: Intel PECI
+    IntelPECI,
+    /// Not all types are supported by all chips So keep space for
+    /// Unknown sensors.
+    Unknown(u8),
+}
+
+impl From<u8> for TermalSensorType {
+    fn from(input: u8) -> Self {
+        match input {
+            0 => Self::CPUEmbeddedDiode,
+            1 => Self::Transistor3904,
+            3 => Self::ThermalDiode,
+            4 => Self::Thermistor,
+            5 => Self::AMDAMDSI,
+            6 => Self::IntelPECI,
+            n => Self::Unknown(n),
+        }
+    }
+}
 
-    if let Ok(dir) = read_dir(folder) {
+/// Check given `item` dispatch to read the right `file` with the right parsing and store data in
+/// given `component`. `id` is provided for `label` creation.
+fn fill_component(component: &mut Component, item: &str, folder: &Path, file: &str) {
+    let hwmon_file = folder.join(file);
+    match item {
+        "type" => {
+            component.sensor_type =
+                read_number_from_file::<u8>(&hwmon_file).map(TermalSensorType::from)
+        }
+        "input" => {
+            let temperature = get_temperature_from_file(&hwmon_file);
+            component.input_file = hwmon_file;
+            component.temperature = temperature;
+            // Maximum know try to get it from `highest` if not available
+            // use current temperature
+            if component.max.is_none() {
+                component.max = temperature;
+            }
+        }
+        "label" => component.label = get_file_line(&hwmon_file, 10).unwrap_or_default(),
+        "highest" => {
+            component.max = get_temperature_from_file(&hwmon_file).or(component.temperature);
+            component.highest_file = Some(hwmon_file);
+        }
+        "max" => component.threshold_max = get_temperature_from_file(&hwmon_file),
+        "min" => component.threshold_min = get_temperature_from_file(&hwmon_file),
+        "crit" => component.threshold_critical = get_temperature_from_file(&hwmon_file),
+        _ => {
+            sysinfo_debug!("This hwmon-temp file is still not supported! Contribution are appreciated.;) {}", hwmon_file)
+        }
+    }
+}
+
+impl Component {
+    /// Read out `hwmon` info (hardware monitor) from `folder` [Path]
+    /// for value path to check at refresh and files containing `max`,
+    /// `critical value` and `label` and store this info into `components`.
+    ///
+    /// Note that a thermal [Component] must have way to read his temperature
+    /// if not it will not be instanciated.
```

Comment:
```suggestion
    /// Note that a thermal [Component] must have a way to read its temperature.
    /// If not, it will be ignored and not added into `components`.
```